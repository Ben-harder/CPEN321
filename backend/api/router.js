import { app } from '../server';
var auth = require('./auth/auth');
var job = require('./job/job');

module.exports = function (app)
{
  // auth routes
  app.post('/auth/create-user', auth.createUser);
  app.get('/auth/user-exists', auth.doesUserExist);
  app.get('/auth/sign-in', auth.userSignIn);

  app.post('/create-job', job.createJob);

  app.post('/applyForJob', (req, res) => {
    res.sendStatus(200);
    console.log("User applied for a job, add job details here.");
  })

  app.get('/get-all-jobs', job.getAllJobs);

  app.get('/get-employer-jobs', job.getEmployerJobs);

  // app.get('/job/get-taken-jobs', job.getTakenJobs);
}
