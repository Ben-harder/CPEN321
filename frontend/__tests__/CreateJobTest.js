import "react-native";
import React from "react";
import CreateJob from "../screens/CreateJobScreen";
import { shallow, configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import NavigationTestUtils from "react-navigation/NavigationTestUtils";
import Expo from "expo";
import configureStore from 'redux-mock-store'
import toJson from 'enzyme-to-json';
import { TextInput } from 'react-native';
import axios from 'axios';
import MockAdapter from 'axios-mock-adapter';
import api from "../constants/Url";

configure({ adapter: new Adapter() });

// mock store
const mockStore = configureStore();
const initialState = {
  user: {
    data: {
      ID: "12345"
    }
  }
};
const store = mockStore(initialState);

// mock expo functions
jest.mock("expo", () => ({
  Constants: {
    manifest: {
      debuggerHost: {
        split: jest.fn().mockReturnValue({shift: jest.fn().mockReturnValue({concat: jest.fn()})})
      }
    }
  }
}));

// mock axios
const mockAxios = new MockAdapter(axios);

// mock alert
global.alert = jest.fn();

// mock navigation
const navigation = {
  navigate: jest.fn()
};

describe("Create Job Test", () => {
  jest.useFakeTimers();
  beforeEach(() => {
    NavigationTestUtils.resetInternalState();
  });

  it("Renders the create job screen", async () => {
    const wrapper = shallow(<CreateJob store={store} />).dive();
    expect(wrapper.instance()).toMatchSnapshot();
  });

  it("Successful job post", async () => {
    // set axios to post to resolve
    mockAxios.onPost(`${api}/job/create-job`).reply(200, {});

    const wrapper = shallow(<CreateJob store={store} navigation={navigation} />).dive();

    // set text inputs
    wrapper.find({testID: '#address'}).simulate('ChangeText', '123 Wain Street');
    wrapper.find({testID: '#description'}).simulate('ChangeText', 'A valid description');
    wrapper.find({testID: '#wage'}).simulate('ChangeText', 20);
    wrapper.find({testID: '#type'}).prop('onValueChange')('Feed Lizard');
    
    // press button
    wrapper.find({testID: '#submit'}).simulate('press');

    const component = wrapper.instance();

    expect(component.state.address === '123 Wain Street');
    expect(component.state.description === 'A valid description');
    expect(component.state.wage === '20');
    expect(component.state.type === 'Feed Lizard');

    // assert success message state
    setTimeout(() => {
      expect(global.alert).toBeCalledWith("You've successfully created the job post.", expect.any(Function))
    }, 0);
  });

  it("Server request errors", async () => {
    // set axios to post to resolve
    mockAxios.onPost(`${api}/job/create-job`).reply(400, {err: {
      response: {
        data: {
          errorMessage: 'Mock error message'
        }
      }
    }});

    const wrapper = shallow(<CreateJob store={store} navigation={navigation} />).dive();

    // set text inputs
    wrapper.find({testID: '#address'}).simulate('ChangeText', '123 Wain Street');
    wrapper.find({testID: '#description'}).simulate('ChangeText', 'A valid description');
    wrapper.find({testID: '#wage'}).simulate('ChangeText', 20);
    wrapper.find({testID: '#type'}).prop('onValueChange')('Feed Lizard');
    
    // press button
    wrapper.find({testID: '#submit'}).simulate('press');

    const component = wrapper.instance();

    expect(component.state.address === '123 Wain Street');
    expect(component.state.description === 'A valid description');
    expect(component.state.wage === '20');
    expect(component.state.type === 'Feed Lizard');

    // assert error message state
    setTimeout(() => {
      expect(global.alert).toBeCalledWith("Mock error message", expect.any(Function))
    }, 0);
  });

  it("Has empty fields", async () => {
    const wrapper = shallow(<CreateJob store={store} navigation={navigation} />).dive();

    // press button
    wrapper.find({testID: '#submit'}).simulate('press');

    // assert error message state
    setTimeout(() => {
      expect(global.alert).toBeCalledWith("Please ensure you've field out the fields.", expect.any(Function))
    }, 0);
  });

  it("Has invalid wage", async () => {
    const wrapper = shallow(<CreateJob store={store} navigation={navigation} />).dive();

    // set text inputs
    wrapper.find({testID: '#address'}).simulate('ChangeText', '123 Wain Street');
    wrapper.find({testID: '#description'}).simulate('ChangeText', 'A valid description');
    wrapper.find({testID: '#wage'}).simulate('ChangeText', "20...0");
    wrapper.find({testID: '#type'}).prop('onValueChange')('Feed Lizard');
    
    // press button
    wrapper.find({testID: '#submit'}).simulate('press');

    const component = wrapper.instance();

    // assert error message state
    setTimeout(() => {
      expect(global.alert).toBeCalledWith("Wage must be a valid format!", expect.any(Function))
    }, 0);
  });

  it("Has invalid address", async () => {
    const wrapper = shallow(<CreateJob store={store} navigation={navigation} />).dive();

    // set text inputs
    wrapper.find({testID: '#address'}).simulate('ChangeText', '123 Wain!!!! Street');
    wrapper.find({testID: '#description'}).simulate('ChangeText', 'A valid description');
    wrapper.find({testID: '#wage'}).simulate('ChangeText', "20");
    wrapper.find({testID: '#type'}).prop('onValueChange')('Feed Lizard');
    
    // press button
    wrapper.find({testID: '#submit'}).simulate('press');

    const component = wrapper.instance();

    // assert error message state
    setTimeout(() => {
      expect(global.alert).toBeCalledWith("Address must be a valid address!", expect.any(Function))
    }, 0);
  });

  it("Component unmounts", async () => {
    const wrapper = shallow(<CreateJob store={store} navigation={navigation} />).dive();

    const component = wrapper.instance();

    component.componentWillUnmount();

    expect(console.disableYellowBox === false);
  });
});
