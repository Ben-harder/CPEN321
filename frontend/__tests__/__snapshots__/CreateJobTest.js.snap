// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Create Job Test Renders the create job screen 1`] = `
CreateJobScreen {
  "attemptCreateJob": [Function],
  "context": Object {},
  "props": Object {
    "dispatch": [Function],
    "store": Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    },
    "storeSubscription": Subscription {
      "listeners": Object {
        "clear": [Function],
        "get": [Function],
        "notify": [Function],
        "subscribe": [Function],
      },
      "onStateChange": [Function],
      "parentSub": undefined,
      "store": Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      },
      "unsubscribe": [Function],
    },
    "user": Object {
      "data": Object {
        "ID": "12345",
      },
    },
  },
  "refs": Object {},
  "setState": [Function],
  "state": Object {
    "address": "",
    "description": "",
    "jobType": "",
    "wage": "",
  },
  "updater": Updater {
    "_callbacks": Array [],
    "_renderer": ReactShallowRenderer {
      "_context": Object {},
      "_element": <CreateJobScreen
        dispatch={[Function]}
        store={
          Object {
            "clearActions": [Function],
            "dispatch": [Function],
            "getActions": [Function],
            "getState": [Function],
            "replaceReducer": [Function],
            "subscribe": [Function],
          }
        }
        storeSubscription={
          Subscription {
            "listeners": Object {
              "clear": [Function],
              "get": [Function],
              "notify": [Function],
              "subscribe": [Function],
            },
            "onStateChange": [Function],
            "parentSub": undefined,
            "store": Object {
              "clearActions": [Function],
              "dispatch": [Function],
              "getActions": [Function],
              "getState": [Function],
              "replaceReducer": [Function],
              "subscribe": [Function],
            },
            "unsubscribe": [Function],
          }
        }
        user={
          Object {
            "data": Object {
              "ID": "12345",
            },
          }
        }
      />,
      "_forcedUpdate": false,
      "_instance": [Circular],
      "_newState": null,
      "_rendered": <TouchableWithoutFeedback
        accessbile={false}
        onPress={[Function]}
      >
        <ScrollViewMock
          style={
            Object {
              "backgroundColor": "#fff",
              "flex": 1,
            }
          }
        >
          <Component
            style={
              Object {
                "alignItems": "center",
              }
            }
          >
            <Text
              accessible={true}
              allowFontScaling={true}
              ellipsizeMode="tail"
              style={
                Object {
                  "fontSize": 30,
                  "paddingBottom": 20,
                }
              }
            >
              Create Job
            </Text>
            <Text
              accessible={true}
              allowFontScaling={true}
              ellipsizeMode="tail"
              style={
                Object {
                  "color": "#62727b",
                  "fontSize": 18,
                  "marginTop": 10,
                  "paddingBottom": 20,
                  "textAlign": "center",
                }
              }
            >
              Please complete the form below to create your job.
            </Text>
            <Text
              accessible={true}
              allowFontScaling={true}
              ellipsizeMode="tail"
              style={
                Object {
                  "color": "#000000",
                  "fontSize": 20,
                  "marginTop": 10,
                  "paddingBottom": 10,
                  "textAlign": "center",
                }
              }
            >
              Address:
            </Text>
            <TextInput
              allowFontScaling={true}
              onChangeText={[Function]}
              placeholder="Enter Address Here"
              returnKeyType="done"
              style={
                Object {
                  "backgroundColor": "#e8e8e8",
                  "borderColor": "#00796b",
                  "borderRadius": 10,
                  "borderWidth": 2,
                  "height": 40,
                  "overflow": "hidden",
                  "padding": 10,
                  "width": "90%",
                }
              }
              testID="#address"
            />
            <Text
              accessible={true}
              allowFontScaling={true}
              ellipsizeMode="tail"
              style={
                Object {
                  "color": "#000000",
                  "fontSize": 20,
                  "marginTop": 10,
                  "paddingBottom": 10,
                  "textAlign": "center",
                }
              }
            >
              Job description:
            </Text>
            <TextInput
              allowFontScaling={true}
              blurOnSubmit={true}
              multiline={true}
              numberOfLines={4}
              onChangeText={[Function]}
              placeholder="Enter Description Here"
              returnKeyType="done"
              style={
                Object {
                  "backgroundColor": "#e8e8e8",
                  "borderColor": "#00796b",
                  "borderRadius": 10,
                  "borderWidth": 2,
                  "height": 100,
                  "overflow": "hidden",
                  "padding": 10,
                  "width": "90%",
                }
              }
              testID="#description"
            />
            <Text
              accessible={true}
              allowFontScaling={true}
              ellipsizeMode="tail"
              style={
                Object {
                  "color": "#000000",
                  "fontSize": 20,
                  "marginTop": 10,
                  "paddingBottom": 10,
                  "textAlign": "center",
                }
              }
            >
              Wage in CAD:
            </Text>
            <TextInput
              allowFontScaling={true}
              keyboardType="numeric"
              onChangeText={[Function]}
              placeholder="Enter Wage here"
              returnKeyType="done"
              style={
                Object {
                  "backgroundColor": "#e8e8e8",
                  "borderColor": "#00796b",
                  "borderRadius": 10,
                  "borderWidth": 2,
                  "height": 40,
                  "overflow": "hidden",
                  "padding": 10,
                  "width": "90%",
                }
              }
              testID="#wage"
            />
            <Text
              accessible={true}
              allowFontScaling={true}
              ellipsizeMode="tail"
              style={
                Object {
                  "color": "#000000",
                  "fontSize": 20,
                  "marginTop": 10,
                  "paddingBottom": 10,
                  "textAlign": "center",
                }
              }
            >
              Job type:
            </Text>
            <RNPickerSelect
              disabled={false}
              doneText="Done"
              hideDoneBar={false}
              hideIcon={false}
              itemKey={null}
              items={
                Array [
                  Object {
                    "label": "Mow Lawn",
                    "value": "Mow Lawn",
                  },
                  Object {
                    "label": "Feed Lizard",
                    "value": "Feed Lizard",
                  },
                  Object {
                    "label": "Do Homework",
                    "value": "Do Homework",
                  },
                  Object {
                    "label": "Clean Pool",
                    "value": "Clean Pool",
                  },
                ]
              }
              modalProps={Object {}}
              onDonePress={null}
              onDownArrow={null}
              onUpArrow={null}
              onValueChange={[Function]}
              pickerProps={Object {}}
              placeholder={
                Object {
                  "label": "     Select Job Here",
                  "value": "",
                }
              }
              placeholderTextColor="#C7C7CD"
              style={Object {}}
              testID="#type"
              textInputProps={Object {}}
            />
            <Component
              style={
                Object {
                  "backgroundColor": "#00796b",
                  "borderColor": "#00796b",
                  "borderRadius": 10,
                  "marginBottom": 10,
                  "overflow": "hidden",
                  "padding": 10,
                }
              }
            >
              <Button
                color="#ffffff"
                onPress={[Function]}
                style={
                  Object {
                    "fontSize": 18,
                    "fontWeight": "600",
                  }
                }
                testID="#submit"
                title="Submit"
              />
            </Component>
          </Component>
        </ScrollViewMock>
      </TouchableWithoutFeedback>,
      "_rendering": false,
      "_updater": [Circular],
    },
  },
  Symbol(enzyme.__setState__): [Function],
}
`;
